### PR TITLE
Fix Issue #2276

### PR DESCRIPTION
**Pull request by yoavain-sundaysky**
Fix Issue #2276

* Convert `source.filePath` to its absolute path to be comparable to importPath/filePath on Windows

🚨 **IMPORTANT: Please do not create a Pull Request without creating an issue first.**

_Any change needs to be discussed before proceeding. Failure to do so may result in the rejection of
the pull request._

## Description

Please include a summary of the change and which issue is fixed. Please also include relevant
motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [X] Bug fix (non-breaking change which fixes an issue)

## Screenshots/Sandbox (if appropriate/relevant):

Adding links to sandbox or providing screenshots can help us understand more about this PR and take
action on it as appropriate

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can
reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

**Test Environment**:

- OS:
- `@graphql-eslint/...`:
- NodeJS:

## Checklist:

- [ ] I have followed the
      [CONTRIBUTING](https://github.com/the-guild-org/Stack/blob/master/CONTRIBUTING.md) doc and the
      style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests and linter rules pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose
the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by yoavain-sundaysky** [comment]
Anyone can review and merge this?

**Comment by yoavain-sundaysky** [comment]
Hey @dotansimha,
Any chance you make this happen?

**Review by dotansimha — APPROVED** [review]

**Review by dimaMachina — COMMENTED** [review]
Hi @yoavain-sundaysky, I just set up running Windows on CI and fixed some issues with Windows. Could you add a test with your change?

**Comment by yoavain** [comment]
> Hi @yoavain-sundaysky, I just set up running Windows on CI and fixed some issues with Windows. Could you add a test with your change?

Sure, I'll do that

**Comment by yoavain-sundaysky** [comment]
Added tests. 
Had to be creative to create a failing test without the fix.
Tests check that it doesn't matter which delimiter is used, the comparison in working

File: packages/plugin/src/rules/require-import-fragment/index.ts [hunk 1]
```diff
@@ -90,13 +90,16 @@ export const rule: GraphQLESLintRule = {
...
           if (!extractedImportPath) continue;
 
-          const importPath = path.join(path.dirname(filePath), extractedImportPath);
+          const importPath = path.join(
+            path.dirname(filePath),
+            extractedImportPath.replaceAll(/[/\\]/g, path.sep),
+          );
+
           const hasInSiblings = fragmentsFromSiblings.some(
             source => source.filePath === importPath,
```
Comment (dimaMachina) [inline packages/plugin/src/rules/require-import-fragment/index.ts:99]: previously you added `path.resolve(source.filePath)` but why?

`source.filePath` comes from `graphql-config` and contains slashes according to user's os

File: packages/plugin/src/rules/require-import-fragment/index.ts [hunk 2]
```diff
@@ -90,13 +90,16 @@ export const rule: GraphQLESLintRule = {
...
           );
           if (hasInSiblings) return;
         }
-
         const fragmentInSameFile = fragmentsFromSiblings.some(
           source => source.filePath === filePath,
```
Comment (dimaMachina) [inline packages/plugin/src/rules/require-import-fragment/index.ts:104]: same, `source.filePath` comes from `graphql-config` and contains slashes according to user's OS, and `filePath` is ESLint's `context.filename` which also contains slashed according user's OS

File: packages/plugin/src/rules/require-import-fragment/index.ts (hunk 1 above)
Comment (yoavain-sundaysky) [reply packages/plugin/src/rules/require-import-fragment/index.ts:94]: I thought so too, but when I printed out the filePath in `fragmentsFromSiblings` by adding 
```
console.log(fragmentsFromSiblings.map(x=>x.filePath));
```

I saw this syntax. On Windows.

```
C:/Dev/my-project/client/logic/experiments/experimentFragment.graphql
```

Comment (dimaMachina) [reply packages/plugin/src/rules/require-import-fragment/index.ts:94]: Strange because I added console log on ci and get backslash on windows

Comment (dimaMachina) [reply packages/plugin/src/rules/require-import-fragment/index.ts:94]: here I commented https://github.com/dimaMachina/graphql-eslint/pull/2277/commits/d9f5b4ea2dd22999fdfa50452a689a7c6e341d8dand got 

IMG_2887

Comment (dimaMachina) [reply packages/plugin/src/rules/require-import-fragment/index.ts:94]: I wonder maybe it happens when you specify glob syntax for `documents`? e.g. `**/*.gql`

Comment (dimaMachina) [reply packages/plugin/src/rules/require-import-fragment/index.ts:94]: yeah, I was right, specifying glob pattern returns forward slash always

image

File: .changeset/empty-horses-relate.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+'@graphql-eslint/eslint-plugin': patch
+---
+
+fix issue #2276
```
Comment (dimaMachina) [inline .changeset/empty-horses-relate.md:5]:
```suggestion
fix false positive cases for `require-import-fragment` on Windows, when `graphql-config`'s `documents` key contained glob pattern => source file path contained always forward slashes
```

File: packages/plugin/src/documents.ts [hunk 4]
```diff
@@ -15,13 +15,18 @@ const handleVirtualPath = (documents: Source[]): Source[] => {
   return documents.map(source => {
     const location = source.location!;
     if (['.gql', '.graphql'].some(extension => location.endsWith(extension))) {
-      return source;
+      return {
+        ...source,
+        // When using glob pattern e.g. `**/*.gql` location contains always forward slashes even on
+        // Windows
+        location: path.resolve(location),
+      };
```
Comment (dimaMachina) [inline packages/plugin/src/documents.ts:23]: @yoavain-sundaysky this is better fix, so other rules will get always os specific slash

**Review by dimaMachina — APPROVED** [review]

**Pull request merged**